File: cpp/inter/stmt_tests.cpp
#include <gtest/gtest.h>

#include "cpp/inter/access.hpp"
#include "cpp/inter/constant.hpp"
#include "cpp/inter/else.hpp"
#include "cpp/inter/identifier.hpp"
#include "cpp/inter/if.hpp"
#include "cpp/inter/node.hpp"
#include "cpp/inter/set.hpp"
#include "cpp/inter/set-elem.hpp"
#include "cpp/inter/statement.hpp"
#include "cpp/inter/statement-sequence.hpp"
#include "cpp/inter/temporary.hpp"

#include "cpp/lexer/token.hpp"
#include "cpp/lexer/word.hpp"
#include "cpp/lexer/type.hpp"

#include <string>
#include <sstream>
#include <vector>

using inter::Access;
using inter::Constant;
using inter::Else;
using inter::Identifier;
using inter::If;
using inter::Node;
using inter::Set;
using inter::SetElem;
using inter::Statement;
using inter::StatementSequence;
using inter::Temporary;

using lexer::Word;
using lexer::Token;

using symbols::Type;

namespace {
struct TestCase {
  std::shared_ptr<Statement> stmt;
  std::string gen;
};
}

TEST(TestStmt, InterTests) {
   std::vector<TestCase> test_cases{
    TestCase{
      Set::create(
        Identifier::create(Word::create("x", Token::kIdentifier), Type::integer, 4),
        Constant::create(42)
      ),
      "\tx = 42\n",
    },
    TestCase{
      SetElem::create(
        Access::create(
          Identifier::create(Word::create("arr", Token::kIdentifier), Type::real, 4),
          Identifier::create(Word::create("x", Token::kIdentifier), Type::integer, 4),
          Type::real
        ),
        Constant::create(42.0)
      ),
      "\tarr[ x ] = 42\n",
    },
    TestCase{
      StatementSequence::create(
        Set::create(
          Identifier::create(Word::create("x", Token::kIdentifier), Type::integer, 4),
          Constant::create(42)
        ),
        SetElem::create(
          Access::create(
            Identifier::create(Word::create("arr", Token::kIdentifier), Type::real, 4),
            Identifier::create(Word::create("x", Token::kIdentifier), Type::integer, 4),
            Type::real
          ),
          Constant::create(42.0)
        )
      ),
      "\tx = 42\nL3:\tarr[ x ] = 42\n",
    },
    TestCase{
      If::create(
        Identifier::create(Word::create("b", Token::kIdentifier), Type::boolean, 4),
        Set::create(
          Identifier::create(Word::create("x", Token::kIdentifier), Type::integer, 4),
          Constant::create(0)
        )
      ),
      "\tiffalse b goto L2\nL3:\tx = 0\n",
    },
    TestCase{
      Else::create(
        Identifier::create(Word::create("b", Token::kIdentifier), Type::boolean, 4),
        Set::create(
          Identifier::create(Word::create("x", Token::kIdentifier), Type::integer, 4),
          Constant::create(0)
        ),
        Set::create(
          Identifier::create(Word::create("x", Token::kIdentifier), Type::integer, 4),
          Constant::create(42)
        )
      ),
      "\tiffalse b goto L4\nL3:\tx = 0\n\tgoto L2\nL4:\tx = 42\n",
    },
   };

   for (auto &tc : test_cases) {
    Node::reset_labels();
    Temporary::reset_temp_count();
    auto begin = Node::new_label();
    auto after = Node::new_label();
    std::stringstream ss;
    tc.stmt->gen(ss, begin, after);
    ASSERT_EQ(ss.str(), tc.gen);
   }
}

File: java/com/dragon/ast/StmtTests.java
package com.dragon.ast;

import static org.junit.Assert.assertEquals;

import java.util.Arrays;

import org.junit.Test;

import com.dragon.lexer.Tag;
import com.dragon.lexer.Word;
import com.dragon.symbols.Type;

public class StmtTests {
  private class TestCase {
    public Stmt stmt;
    public String gen;

    public TestCase(Stmt stmt, String gen) {
      this.stmt = stmt;
      this.gen = gen;
    }
  }

  @Test
  public void stmtTests() {
    TestCase[] testCases = {
      new TestCase(
        new Set(
          new Id(new Word("x", Tag.ID), Type.Int, 4),
          new Constant(42)
        ),
        "\tx = 42\n"
      ),
      new TestCase(
        new SetElem(
          new Access(
            new Id(new Word("arr", Tag.ID), Type.Float, 4),
            new Id(new Word("x", Tag.ID), Type.Int, 4),
            Type.Float),
          new Constant(42.0)
        ),
        // TODO(sambatyon): Formatting of floats is using PC locale.
        "\tarr [ x ] = 42,000\n"
      ),
      new TestCase(
        new Seq(
          new Set(
            new Id(new Word("x", Tag.ID), Type.Int, 4),
            new Constant(42)
          ),
          new SetElem(
            new Access(
              new Id(new Word("arr", Tag.ID), Type.Float, 4),
              new Id(new Word("x", Tag.ID), Type.Int, 4),
              Type.Float),
            new Constant(42.0)
          )
        ),
        "\tx = 42\nL3:\tarr [ x ] = 42,000\n"
      ),
      new TestCase(
        new If(
          new Id(new Word("b", Tag.ID), Type.Bool, 4),
          new Set(
            new Id(new Word("x", Tag.ID), Type.Int, 4),
            new Constant(0)
          )
        ),
        "\tiffalse b goto L2\nL3:\tx = 0\n"
      ),
      new TestCase(
        new Else(
          new Id(new Word("b", Tag.ID), Type.Bool, 4),
          new Set(
            new Id(new Word("x", Tag.ID), Type.Int, 4),
            new Constant(0)
          ),
          new Set(
            new Id(new Word("x", Tag.ID), Type.Int, 4),
            new Constant(42)
          )
        ),
        "\tiffalse b goto L4\nL3:\tx = 0\n\tgoto L2\nL4:\tx = 42\n"
      ),
      new TestCase(
        new While(
          new Id(new Word("b", Tag.ID), Type.Bool, 4),
          new Set(
            new Id(new Word("x", Tag.ID), Type.Int, 4),
            new Constant(0)
          )
        ),
        "\tiffalse b goto L2\nL3:\tx = 0\n\tgoto L1\n"
      ),
      new TestCase(
        new Do(
          new Set(
            new Id(new Word("x", Tag.ID), Type.Int, 4),
            new Constant(0)
          ),
          new Id(new Word("b", Tag.ID), Type.Bool, 4)
        ),
        "\tx = 0\nL3:\tif b goto L1\n"
      ),
    };

    for (var tc : Arrays.asList(testCases)) {
      Node.resetLabel();
      Temp.resetTempCount();

      var begin = Node.newLabel();
      var after = Node.newLabel();
      var builder = new StringBuilder();

      tc.stmt.gen(builder, begin, after);
      assertEquals(builder.toString(), tc.gen);
    }
  }
}
